MTL-2348 CSM 1.4.4 Patch (#4647)
Cray-HPE/docs-csm: upgrade/1.4.4/CSM-Only.md

# CSM Only Upgrade

> This page provides guidance exclusively for systems upgrading CSM v1.4.X to CSM v1.4.4.

Reviewer: ```suggestion
This page provides guidance for systems with exclusively CSM installed that are performing an upgrade from version CSM v1.4.X to CSM v1.4.4.
```

> The [v1.4.4](../1.4.4/README.md) upgrade page will refer to this page

Reviewer 2: ```suggestion
The [v1.4.4 upgrade page](../1.4.4/README.md) will refer to this page
```

during [Update NCN images](../1.4.4/README.md#update-ncn-images).

**If other products are installed on the system, return to [Update NCN images](../1.4.4/README.md#update-ncn-images) and
choose option 2.**

## Requirements

> * `CSM_RELEASE` is set in the shell environment on `ncn-m001`.

Reviewer 2: ```suggestion
* `CSM_RELEASE` is set in the shell environment on `ncn-m001`. See [Preparation](README.md#preparation) for details on how it should be set.
```

## Steps

> 1. (`ncn-m001#`) Generate a new CFS configuration for the management nodes.

Reviewer 2: ```suggestion
1. (`ncn-m001#`) Generate a new [CFS](../../glossary.md#configuration-framework-service-cfs) configuration for the management nodes.
```

   This script creates a new CFS configuration that includes the CSM version in its name and applies it to the
   management nodes. This leaves the management node components in CFS disabled. They will be automatically enabled when
   they are rebooted at a later stage in the upgrade.

   ```bash
   /usr/share/doc/csm/scripts/operations/configuration/apply_csm_configuration.sh \
       --no-enable --config-name "management-${CSM_RELEASE}"
   ```

   Successful output should end with the following line:

   ```text
   All components updated successfully.
   ```

> 1. Set the IMS IDs for each Management node sub-role.

Reviewer 2: ```suggestion
1. (`ncn-m001#`) Set the [IMS](../../glossary.md#image-management-service-ims) IDs for each Management node sub-role.
```

    * Kubernetes ID

      ```bash
      KUBERNETES_IMAGE_ID="$(kubectl -n services get cm cray-product-catalog -o jsonpath='{.data.csm}' \
        | yq r -j - '"'${CSM_RELEASE}'".images' \
        | jq -r '. as $o | keys_unsorted[] | select(startswith("secure-kubernetes")) | $o[.].id')"
      export KUBERNETES_IMAGE_ID
      echo "KUBERNETES_IMAGE_ID=$KUBERNETES_IMAGE_ID"
      ```

    * Storage-CEPH ID

      ```bash
      STORAGE_IMAGE_ID="$(kubectl -n services get cm cray-product-catalog -o jsonpath='{.data.csm}' \
        | yq r -j - '"'${CSM_RELEASE}'".images' \
        | jq -r '. as $o | keys_unsorted[] | select(startswith("secure-storage")) | $o[.].id')"
      export STORAGE_IMAGE_ID
      echo "STORAGE_IMAGE_ID=$STORAGE_IMAGE_ID"
      ```

> 1. Create new CFS sessions for both Kubernetes and Storage nodes.

Reviewer 2: ```suggestion
1. (`ncn-m001#`) Create CFS sessions for both Kubernetes and Storage nodes.
```

> 1. Create new CFS sessions for both Kubernetes and Storage nodes.

Reviewer 2: You may want to have the session names include a timestamp. Or else include a note that these commands will fail if being run for a second time, because the CFS sessions already exist (in which case they would first need to be deleted). You could just save a timestamp to an environment variable, use it for both session names, so that you can then use it when querying the session statuses in the subsequent command.

Reviewer 2: I added some additional comments with suggested changes to address this

>     1. Build master images.

Reviewer 2: ```suggestion
    1. Build customized master node images.
```

       ```bash
       cray cfs sessions create \
           --target-group Management_Master \
           "$KUBERNETES_IMAGE_ID" \
           --target-definition image \
           --target-image-map "$KUBERNETES_IMAGE_ID" "" \
           --configuration-name "management-${CSM_RELEASE}" \
           --name "management-kubernetes-${CSM_RELEASE}-upgrade" \
           --format json
       ```

>     1. Build storage images.

Reviewer 2: ```suggestion
    1. Build customized storage node images.
```

       ```bash
       cray cfs sessions create \
           --target-group Management_Storage \
           "$STORAGE_IMAGE_ID" \
           --target-definition image \

>            --target-image-map "$STORAGE_IMAGE_ID" "" \

Reviewer 2: If you don't care what name the resulting images have in IMS, you can just omit the --target-image-map parameter entirely. That just defines what the "name" field will have for the IMS image. 

Alternatively, you could set it to something like descriptive that would give people an idea of what it is if they were just listing IMS images.

```suggestion
           --target-image-map "$KUBERNETES_IMAGE_ID" "management-kubernetes-${CSM_RELEASE}" \
           --configuration-name "management-${CSM_RELEASE}" \
           --name "management-kubernetes-${CSM_RELEASE}-upgrade" \
           --format json
       ```

    1. Build customized storage node images.

       ```bash
       cray cfs sessions create \
           --target-group Management_Storage \
           "$STORAGE_IMAGE_ID" \
           --target-definition image \
           --target-image-map "$STORAGE_IMAGE_ID" "management-storage-${CSM_RELEASE}" \
```

           --configuration-name "management-${CSM_RELEASE}" \
           --name "management-storage-${CSM_RELEASE}-upgrade" \
           --format json
       ```

> 1. Wait for the image builds to complete successfully

Reviewer 2: ```suggestion
1. Wait for the image builds to complete successfully. This may take over 20 minutes to complete.
```

> 1. Wait for the image builds to complete successfully
> 
>    ```bash

Reviewer 2: ```suggestion
1. (`ncn-m001#`) Wait for the image builds to complete successfully.

   > This may take over 20 minutes to complete.

   ```bash
```

   watch '
   cray cfs sessions describe "management-kubernetes-${CSM_RELEASE}-upgrade" --format json | jq -r ".status.session.succeeded"
   cray cfs sessions describe "management-storage-${CSM_RELEASE}-upgrade" --format json | jq -r ".status.session.succeeded"
   '
   ```

   Expected results:

   ```text
   true
   true
   ```

> 1. Set the image ID

Reviewer 2: ```suggestion
1. (`ncn-m001#`) Set environment variables with the IMS image IDs.
```

>     * Kubernetes ID

Reviewer 2: ```suggestion
    * Kubernetes image IMS ID
```

      ```bash
      NEW_KUBERNETES_IMAGE_ID="$(cray cfs sessions describe "management-kubernetes-${CSM_RELEASE}-upgrade" --format json | jq -r '.status.artifacts[].image_id')"
      export NEW_KUBERNETES_IMAGE_ID
      echo "NEW_KUBERNETES_IMAGE_ID=$NEW_KUBERNETES_IMAGE_ID"
      ```

>     * Storage-CEPH

Reviewer 2: ```suggestion
    * Storage image IMS ID
```

      ```bash
      NEW_STORAGE_IMAGE_ID="$(cray cfs sessions describe "management-storage-${CSM_RELEASE}-upgrade" --format json | jq -r '.status.artifacts[].image_id')"
      export NEW_STORAGE_IMAGE_ID
      echo "NEW_STORAGE_IMAGE_ID=$NEW_STORAGE_IMAGE_ID"
      ```

> 1. Assign images in BSS

Reviewer 2: ```suggestion
1. (`ncn-m001#`) Assign images in [BSS](../../glossary.md#boot-script-service-bss).
```

# CSM Only Upgrade

> This page provides guidance for systems with exclusively CSM installed that are performing an upgrade from version CSM v1.4.X to CSM v1.4.4.

Reviewer 2: ```suggestion
This page provides guidance for systems with exclusively CSM installed that are performing an upgrade from a CSM v1.4.X release to the CSM v1.4.4 release.
```

       --no-enable --config-name "management-${CSM_RELEASE}"
   ```

   Successful output should end with the following line:

   ```text
   All components updated successfully.
   ```

1. (`ncn-m001#`) Set the [IMS](../../glossary.md#image-management-service-ims) IDs for each Management node sub-role.

    * Kubernetes ID

      ```bash
      KUBERNETES_IMAGE_ID="$(kubectl -n services get cm cray-product-catalog -o jsonpath='{.data.csm}' \
        | yq r -j - '"'${CSM_RELEASE}'".images' \
        | jq -r '. as $o | keys_unsorted[] | select(startswith("secure-kubernetes")) | $o[.].id')"
      export KUBERNETES_IMAGE_ID
      echo "KUBERNETES_IMAGE_ID=$KUBERNETES_IMAGE_ID"
      ```

    * Storage-CEPH ID

      ```bash
      STORAGE_IMAGE_ID="$(kubectl -n services get cm cray-product-catalog -o jsonpath='{.data.csm}' \
        | yq r -j - '"'${CSM_RELEASE}'".images' \
        | jq -r '. as $o | keys_unsorted[] | select(startswith("secure-storage")) | $o[.].id')"
      export STORAGE_IMAGE_ID
      echo "STORAGE_IMAGE_ID=$STORAGE_IMAGE_ID"
      ```

1. (`ncn-m001#`) Create CFS sessions for both Kubernetes and Storage nodes.

>     1. Build customized master node images.
> 
>        ```bash
>        cray cfs sessions create \
>            --target-group Management_Master \
>            "$KUBERNETES_IMAGE_ID" \
>            --target-definition image \
>            --target-image-map "$KUBERNETES_IMAGE_ID" "management-kubernetes-${CSM_RELEASE}" \
>            --configuration-name "management-${CSM_RELEASE}" \
>            --name "management-kubernetes-${CSM_RELEASE}-upgrade" \
>            --format json
>        ```
> 
>     1. Build customized storage node images.
> 
>        ```bash
>        cray cfs sessions create \
>            --target-group Management_Storage \
>            "$STORAGE_IMAGE_ID" \
>            --target-definition image \
>            --target-image-map "$STORAGE_IMAGE_ID" "management-storage-${CSM_RELEASE}" \
>            --configuration-name "management-${CSM_RELEASE}" \
>            --name "management-storage-${CSM_RELEASE}-upgrade" \
>            --format json
>        ```
> 
> 1. (`ncn-m001#`) Wait for the image builds to complete successfully.
> 
>    > This may take over 20 minutes to complete.
> 
>    ```bash
>    watch '
>    cray cfs sessions describe "management-kubernetes-${CSM_RELEASE}-upgrade" --format json | jq -r ".status.session.succeeded"
>    cray cfs sessions describe "management-storage-${CSM_RELEASE}-upgrade" --format json | jq -r ".status.session.succeeded"

Reviewer 2: ```suggestion
    1. Set variables with unique names for the CFS sessions to be created.

       ```bash
       TSTAMP=$(date "+%Y%m%d%H%M%S")
       K8S_CFS_SESSION_NAME="management-kubernetes-${CSM_RELEASE}-${TSTAMP}"
       STOR_CFS_SESSION_NAME="management-storage-${CSM_RELEASE}-${TSTAMP}"
       ```

    1. Build customized master node images.

       ```bash
       cray cfs sessions create \
           --target-group Management_Master \
           "$KUBERNETES_IMAGE_ID" \
           --target-definition image \
           --target-image-map "$KUBERNETES_IMAGE_ID" "management-kubernetes-${CSM_RELEASE}" \
           --configuration-name "management-${CSM_RELEASE}" \
           --name "${K8S_CFS_SESSION_NAME}" \
           --format json
       ```

    1. Build customized storage node images.

       ```bash
       cray cfs sessions create \
           --target-group Management_Storage \
           "$STORAGE_IMAGE_ID" \
           --target-definition image \
           --target-image-map "$STORAGE_IMAGE_ID" "management-storage-${CSM_RELEASE}" \
           --configuration-name "management-${CSM_RELEASE}" \
           --name "${STOR_CFS_SESSION_NAME}" \
           --format json
       ```

1. (`ncn-m001#`) Wait for the image builds to complete successfully.

   > This may take over 20 minutes to complete.

   ```bash
   watch '
   cray cfs sessions describe "${K8S_CFS_SESSION_NAME}" --format json | jq -r ".status.session.succeeded"
   cray cfs sessions describe "${STOR_CFS_SESSION_NAME}" --format json | jq -r ".status.session.succeeded"
```

   '
   ```

   Expected results:

   ```text
   true
   true
   ```

1. (`ncn-m001#`) Set environment variables with the IMS image IDs.

    * Kubernetes image IMS ID

      ```bash

>       NEW_KUBERNETES_IMAGE_ID="$(cray cfs sessions describe "management-kubernetes-${CSM_RELEASE}-upgrade" --format json | jq -r '.status.artifacts[].image_id')"

Reviewer 2: ```suggestion
      NEW_KUBERNETES_IMAGE_ID="$(cray cfs sessions describe "${K8S_CFS_SESSION_NAME}" --format json | jq -r '.status.artifacts[].image_id')"
```

      export NEW_KUBERNETES_IMAGE_ID
      echo "NEW_KUBERNETES_IMAGE_ID=$NEW_KUBERNETES_IMAGE_ID"
      ```

    * Storage image IMS ID

      ```bash

>       NEW_STORAGE_IMAGE_ID="$(cray cfs sessions describe "management-storage-${CSM_RELEASE}-upgrade" --format json | jq -r '.status.artifacts[].image_id')"

Reviewer 2: ```suggestion
      NEW_STORAGE_IMAGE_ID="$(cray cfs sessions describe "${STOR_CFS_SESSION_NAME}" --format json | jq -r '.status.artifacts[].image_id')"
```

# CSM Only Upgrade

This page provides guidance exclusively for systems upgrading CSM v1.4.X to CSM v1.4.4.

The [v1.4.4](../1.4.4/README.md) upgrade page will refer to this page
during [Update NCN images](../1.4.4/README.md#update-ncn-images).

**If other products are installed on the system, return to [Update NCN images](../1.4.4/README.md#update-ncn-images) and
choose option 2.**

## Requirements

* `CSM_RELEASE` is set in the shell environment on `ncn-m001`.

## Steps

1. (`ncn-m001#`) Generate a new CFS configuration for the management nodes.

   This script creates a new CFS configuration that includes the CSM version in its name and applies it to the
   management nodes. This leaves the management node components in CFS disabled. They will be automatically enabled when
   they are rebooted at a later stage in the upgrade.

   ```bash
   /usr/share/doc/csm/scripts/operations/configuration/apply_csm_configuration.sh \
       --no-enable --config-name management-${CSM_RELEASE}
   ```

   Successful output should end with the following line:

   ```text
   All components updated successfully.
   ```

> 1. Set the IMS IDs for each Management node subrole.

Author: I need a pragmatic way to get these. I know I can get the IMS ID from `kubectl -n services get cm cray-product-catalog -o jsonpath='{.data.csm}'` but the image name has a non-deterministic path.

```bash
ncn-m001:~ # kubectl -n services get cm cray-product-catalog -o jsonpath='{.data.csm}' | yq4 eval '."1.4.4-beta.1".images' -
cray-shasta-csm-sles15sp4-barebones.x86_64-csm-1.4:
  id: 2648b5d9-8a49-4164-9298-125249ae7d61
secure-kubernetes-0.4.71-x86_64.squashfs:
  id: 9574bcd8-aef5-449d-858e-c0ed36a20816
secure-storage-ceph-0.3.59-x86_64.squashfs:
  id: a68077ac-7d6f-4fe3-ab6f-58f25bcff068
```

I can’t grab those without knowing the secure-kubernetes-0.4.71-x86_64.squashfs value

Reviewer 3: Here's one way (example from mug, so version is different):

```
ncn-m001:~ # kubectl -n services get configmap cray-product-catalog -o jsonpath='{.data.csm}' | yq r -j - | jq -r '.["1.5.0-beta.22"].images | to_entries | map(select(.key | contains("secure-kubernetes")))[].value.id'
545ba744-d5e4-4725-9960-3bbe2d9be8d0
```

   ```bash
   export MASTER_IMAGE_ID=""
   ```

   ```bash
   export WORKER_IMAGE_ID=""
   ```

   ```bash
   export STORAGE_IMAGE_ID=""
   ```

1. Create new CFS sessions for both Kubernetes and Storage nodes.

    1. Build master images.

       ```bash
       cray cfs sessions create \
           --target-group Management_Master \
           "$MASTER_IMAGE_ID}" \
           --target-definition image \
           --target-image-map "$MASTER_IMAGE_ID" "" \
           --configuration-name "management-${CSM_RELEASE}" \
           --name "management-master-${CSM_RELEASE}-upgrade" \
           --format json

>        ```

Author: Do we need this at all? For CSM only systems, the master and worker images are the same.

    1. Build storage images.

       ```bash
       cray cfs sessions create \
           --target-group Management_Storage \
           "$STORAGE_IMAGE_ID" \
           --target-definition image \
           --target-image-map "$STORAGE_IMAGE_ID" "" \
           --configuration-name "management-${CSM_RELEASE}" \
           --name "management-storage-${CSM_RELEASE}-upgrade" \
           --format json
       ```

    1. Build worker images.

       ```bash
       cray cfs sessions create \
           --target-group Management_Worker \
           "$WORKER_IMAGE_ID" \
           --target-definition image \
           --target-image-map"$WORKER_IMAGE_ID" "" \
           --configuration-name "management-${CSM_RELEASE}" \
           --name "management-worker-${CSM_RELEASE}-upgrade" \
           --format json
       ```

> 1. Wait forever for images to build.

Author: TODO: Replace this with something that is for checking the command.

When I wrote this, my machine (surtur)'s craycli was broken and I couldn't form any commands to test before signing off.

> 1. Set the new IMS Image IDs

Author: [Same comment](https://github.com/Cray-HPE/docs-csm/pull/4647/files#r1476874723).

   ```bash
   export NEW_MASTER_IMAGE_ID=""
   ```

   ```bash
   export NEW_WORKER_IMAGE_ID=""
   ```

   ```bash
   export NEW_STORAGE_IMAGE_ID=""
   ```

1. Assign images in BSS

    1. Update master management nodes:

       ```bash
       /usr/share/doc/csm/scripts/operations/configuration/node_management/assign-ncn-images.sh \
           -m \
           -p "$NEW_MASTER_IMAGE_ID"

>        ```

Author: [Same comment](https://github.com/Cray-HPE/docs-csm/pull/4647/files#r1476874947), but instead of removing this I could just get both Masters and Workers with:

```bash
       /usr/share/doc/csm/scripts/operations/configuration/node_management/assign-ncn-images.sh \
           -mw \
           -p "$NEW_KUBERNETES_IMAGE_ID"
```

Author: Resolved, using `-mw`.
